crates/pessimistic-proof/src/certificate.rs

```diff
@@ -1,41 +1,149 @@
+    bridge_exit::BridgeExit,
+    imported_bridge_exit::ImportedBridgeExit,
+    keccak::Digest,
+    multi_batch_header::MultiBatchHeader,
+    nullifier_tree::{NullifierKey, NullifierPath},
+};
+use crate::{bridge_exit::TokenInfo, local_balance_tree::LocalBalancePath};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Conversion error: {0}")]
+    TypeConversion(&'static str),
+}
+
+pub type Height = u64;
 
-/// Represents the data submitted by the CDKs to the AggLayer.
+/// Represents the data submitted by the chains to the AggLayer.
 ///
 /// The bridge exits plus the imported bridge exits define
-/// the state transition, resp. the amount that goes out and the amount that comes in.
+/// the state transition, resp. the amount that goes out and the amount that
+/// comes in.
 ///
-/// The bridge exits refer to the [`BridgeExit`]  emitted by
+/// The bridge exits refer to the [`BridgeExit`] emitted by
 /// the origin network of the [`Certificate`].
 ///
 /// The imported bridge exits refer to the [`BridgeExit`] received and imported
 /// by the origin network of the [`Certificate`].
-#[derive(Default, Clone, Debug, Serialize, Deserialize)]
+#[derive(Default, Serialize, Deserialize, Clone, Debug)]
 pub struct Certificate {
-    /// The origin network which emitted this certificate.
-    pub origin_network: NetworkId,
-    /// The initial local exit root.
+    pub network_id: NetworkId,
+    pub height: Height,
     pub prev_local_exit_root: Digest,
-    /// The set of bridge exits emitted by the origin network.
+    pub new_local_exit_root: Digest,
     pub bridge_exits: Vec<BridgeExit>,
-    /// The set of imported bridge exits for which the origin network is the target.
-    pub imported_bridge_exits: Vec<BridgeExit>,
+    pub imported_bridge_exits: Vec<ImportedBridgeExit>,
+    pub signature: Signature,
 }
 
 impl Certificate {
-    /// Creates a new [`Certificate`].
-    pub fn new(
-        origin_network: NetworkId,
-        prev_local_exit_root: Digest,
-        bridge_exits: Vec<BridgeExit>,
-    ) -> Self {
+    pub fn new_for_test(network_id: NetworkId, height: Height) -> Self {
         Self {
-            origin_network,
-            prev_local_exit_root,
-            bridge_exits,
-            imported_bridge_exits: Default::default(),
+            network_id,
+            height,
+            prev_local_exit_root: [0; 32],
+            new_local_exit_root: [0; 32],
+            bridge_exits: Vec::new(),
+            imported_bridge_exits: Vec::new(),
+            signature: Signature {
+                r: U256::ZERO,
+                s: U256::ZERO,
+                odd_y_parity: false,
+            },
         }
     }
+
+    pub fn hash(&self) -> Digest {
+        keccak256_combine([
+            self.network_id.to_be_bytes().as_slice(),
+            self.height.to_be_bytes().as_slice(),
+            self.prev_local_exit_root.as_slice(),
+            self.new_local_exit_root.as_slice(),
+        ])
+    }
```

Comment:
It may be useful to add a comment up where the `Certificate` struct is defined that anyone who adds or modifies the fields there should think whether `Certificate::hash` also needs updating. Seems it could be easily forgotten.

Even better, factor out the parts that are committed to into a separate struct and define the hash as just the hash of a serialization of the inner struct. That way, adding a field to the new struct is automatically picked up when hashing. I am not sure about possible drawbacks of this approach in terms of code clarity and so on but may be worth considering.

Reply:
> I am not sure about possible drawbacks of this approach in terms of code clarity and so on but may be worth considering.

It touches the most commonly accessed fields of this struct, so having some kind of `certificate.inner.foo` whenever we access them might not be so great no?

The access could be behind methods like `network_id()` `prev_local_exit_root()` etc but not sure either how it would impact the code clarity on some area

In any case, I added a note in the `Certificate` rustdoc